### PR TITLE
convert to typescript

### LINKED ISSUE #73: convert to typescript
Hi,

after having a frustrating experience with many CI/CD systems I feel like cirrus ci might be the middle ground I'm looking for, at least backend functionality-wise.
However I feel like the UI / UX leaves a lot to be desired and I also think the UI is quite confusing at times.
I understand you are a small team or just one person company? and you have to prioritize accordingly and perhaps the focus is currently more on backend stuff.
I honestly thought a bit that I could just build a custom UI for cirrus for internal use in my company but positively realized that your UI is OSS.
So that brings me to that idea: Maybe I should contribute a bit. I have quite some experience with React, graphql etc. but I'm kinda opionated about using TypeScript for almost everything these days, especially for side projects:-). Therefore I'd like to ask if you're open to a PR which converts this project to TypeScript. In the context of this project I think it also makes OSS contributions easier because the codebase is easier to understand / navigate. Yeah I could give you a huge list of TypeScript is saving the world, but I'd rather leave that to other folks and their Medium posts ;-)

So: Would you be open to an PR which turns this into TypeScript?

### PR DESCRIPTION
**Pull request by geekflyer**
This is now ready for review.

In summary: This converts the project to TypeScript.

A few details about the changes:
- Currently the settings in `tsconfig.json` are relatively loose / permissive. For example usually I would enable `strictNullChecks: true` but there are about 70 more static type errors that pop up when I enable this. May resolve this in another PR if I find the motivation later. Same is true with `noImplicitAny: true` which I would normally set. TLDR: This PR just is a first step to convert this to TypeScript, but it's not quite best practice.
- As you can see there are a couple of places where I had to explicitly cast things to `any`. This was either because I was too lazy or because I couldn't figure out from the existing JS code what the shape of the variables actually is or because there was technically a bug in the app that TypeScript detected but that I didn't want to fix as part of this commit. For example there are a bunch of cases where a parent component is passing some data of it's own fragment to a child component and TypeScript complains about type incompatibility due a missing fragmentRef. I had to cast those fragment properties to `any` in those cases, however as far as I can tell this is actually a bug in the app (that can also be seen in the browser console). Basically in those cases usually the parent fragment / query does not explicitly include the fragment that the child requires (i.e. a `...ChildFragment_foo` is missing). One such example is a missing `...RepositoryNameChip_repository` in `LastDefaultBranchBuildMiniRow.tsx`. Should probably fix those queries in a follow-up commit.
- There are a few semantic changes I made as well. In particular there were a bunch of places where a styles / css property was set that didn't exist or had no effect. Also there were a bunch of places where a `tooltip` property was set that didn't exist on the material ui component at all. I replaced those with a proper `<Tooltip>` component which actually now properly shows tooltips in places like the "repository settings button" etc. Also the Sign In button now shows a github symbol. All of the semantic changes I made were when typescript was yelling at me and I figured this is actually a "bug" in the code that I can resolve myself. In other words TypeScript already helped to find a few minor bugs. Also introduced a small utility function `getNodesFromConnection` to reduce some boilerplate when turning a connection into an array and to make this conversion typesafe.

closes #73

### CODE REVIEW

**Comment by geekflyer** [comment]
this is now ready for review @fkorotkov @RDIL

**Comment by fkorotkov** [comment]
Awesome work! I'm finishing work on UI for https://github.com/cirruslabs/cirrus-ci-docs/issues/326 and will try to rebase and convert your PR to taste Typescript and do a proper review. 🙌

**Comment by geekflyer** [comment]
do you have an estimate on when this may get merged?
I wanna create some small enhancements, but want to base my work off a master which has this already merged in.

**Comment by fkorotkov** [comment]
@geekflyer I've started looking into it after #85. Will rebase and convert the remainings to typescript as an exercise.

**Comment by fkorotkov** [comment]
I've looked through all the files and everything looks great! Thank you for doing this! I've noticed a few things that I want to investigate like types in `QueryRenderer`s and adding types to utils. Will do this first and then will rebase to convert metrics. I expect to merge it tomorrow. 🙌 Thanks again for championing it! 🎉

File: src/components/TaskCommandLogs.tsx [hunk 1]
```diff
@@ -132,7 +136,9 @@ class TaskCommandLogs extends React.Component<TaskCommandLogsProps> {
               </div>
             );
           }
-          return <TaskCommandRealTimeLogs initialLogLines={props.task.commandLogsTail || []} {...this.props} />;
+          return (
+            <TaskCommandRealTimeLogs initialLogLines={(props.task.commandLogsTail || []).concat()} {...this.props} />
```
Comment (geekflyer) [inline src/components/TaskCommandLogs.tsx]: what's the empty .concat() for?

Comment (fkorotkov) [reply src/components/TaskCommandLogs.tsx]: To convert from `readonly strring[]` to just `string[]` 🤷‍♂

**Comment by fkorotkov** [comment]
Tomorrow I'll go though a few more places with `any` and probably will add minimum type coverage check via https://github.com/plantain-00/type-coverage. Then will rebase, convert and merge! 🙌

File: src/components/TaskCommandLogs.tsx (hunk 1 above)
Comment (geekflyer) [reply src/components/TaskCommandLogs.tsx]: 😱

**Pull request merged**